File: 98-validate-binary-search-tree/validate-binary-search-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isValidBST(root: TreeNode | null): boolean {
    const is_valid = (head:TreeNode, min, max):boolean => {
        if(head === null) {
            return true
        } 
        if (head.val > min && head.val < max) {
            const left:boolean = is_valid(head.left, min, head.val);
            const right:boolean = is_valid(head.right, head.val, max)
            return left && right
        } else return false
    } 
    return is_valid(root, Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER)
};

File: 98-validate-binary-search-tree/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def is_valid(root:Optional[TreeNode], n:int, x:int) -> bool:
            if root == None:
                return True
            
            if (root.val > n and root.val < x):
                left:bool = is_valid(root.left, n, root.val)
                right:bool = is_valid(root.right, root.val, x)
                return left and right
            else: return False

        return is_valid(root, -sys.maxsize-1, sys.maxsize)
            